tls/extensions/s2n_extension_type_lists.c

```diff
@@ -0,0 +1,139 @@
+#include "tls/extensions/s2n_server_session_ticket.h"
+#include "tls/extensions/s2n_server_server_name.h"
+#include "tls/extensions/s2n_server_signature_algorithms.h"
+#include "tls/extensions/s2n_server_supported_versions.h"
+#include "tls/extensions/s2n_server_key_share.h"
+
+static const s2n_extension_type *const client_hello_extensions[] = {
+        &s2n_client_supported_versions_extension,
+        &s2n_client_key_share_extension,
+        &s2n_client_signature_algorithms_extension,
+        &s2n_client_server_name_extension,
+        &s2n_client_alpn_extension,
+        &s2n_client_status_request_extension,
+        &s2n_client_sct_list_extension,
+        &s2n_client_max_frag_len_extension,
+        &s2n_client_session_ticket_extension,
+        &s2n_client_supported_groups_extension,
+        &s2n_client_ec_point_format_extension,
+        &s2n_client_pq_kem_extension,
+        &s2n_client_renegotiation_info_extension,
+};
+
+static const s2n_extension_type *const tls12_server_hello_extensions[] = {
+        &s2n_server_supported_versions_extension,
+        &s2n_server_server_name_extension,
+        &s2n_server_ec_point_format_extension,
+        &s2n_server_renegotiation_info_extension,
+        &s2n_server_alpn_extension,
+        &s2n_server_status_request_extension,
+        &s2n_server_sct_list_extension,
+        &s2n_server_max_fragment_length_extension,
+        &s2n_server_session_ticket_extension,
+};
+
+static const s2n_extension_type *const tls13_server_hello_extensions[] = {
+        &s2n_server_supported_versions_extension,
+        &s2n_server_key_share_extension,
+        &s2n_server_cookie_extension,
+};
+
+static const s2n_extension_type *const encrypted_extensions[] = {
+        &s2n_server_server_name_extension,
+        &s2n_server_max_fragment_length_extension,
+        &s2n_server_alpn_extension,
+};
+
+static const s2n_extension_type *const cert_req_extensions[] = {
+        &s2n_server_signature_algorithms_extension,
+};
+
+static const s2n_extension_type *const certificate_extensions[] = {
+        &s2n_tls13_server_status_request_extension,
+        &s2n_server_sct_list_extension,
+};
+
+#define S2N_EXTENSION_LIST(list) { .extension_types = (list), .count = s2n_array_len(list) }
+#define S2N_EMPTY_EXTENSION_LIST { .extension_types = NULL, .count = 0 }
+
+static struct {
+    s2n_extension_type_list default_list;
+    s2n_extension_type_list tls13_list; /* TLS1.3 moved some extensions to different extension lists */
+} extension_lists[] = {
+        [S2N_EXTENSION_LIST_CLIENT_HELLO] = {
+                .default_list = S2N_EXTENSION_LIST(client_hello_extensions),
+                .tls13_list = S2N_EXTENSION_LIST(client_hello_extensions),
+        },
+        [S2N_EXTENSION_LIST_SERVER_HELLO] = {
+                .default_list = S2N_EXTENSION_LIST(tls12_server_hello_extensions),
+                .tls13_list = S2N_EXTENSION_LIST(tls13_server_hello_extensions),
+        },
```

Comment:
Q: What would this list look like for server hello retry?  Is it the same as `tls13_server_hello_extensions`?

Reply:
At the moment it is, because they both use the same method in our current logic: `s2n_server_extensions_send.` But we could decide later to use a different list to enforce what extensions are and aren't allowed.